### DISCUSSION TITLE
KEDA Operator Get All Access

### DISCUSSION THREAD

**Discussion by alexanderkevin** [kedacore/keda#4730] [Q&A / Need Help]
Hi All,

I want to do a quick check before using KEDA on production. Notice that the Cluster role for KEDA Operator has given get access to all api groups and resources
https://github.com/kedacore/charts/blob/main/keda/templates/manager/clusterrole.yaml#L41-L46
```
- apiGroups:
  - '*'
  resources:
  - '*'
  verbs:
  - get
 ```
is there any particular reason for this or do we have a list of mandatory GET access that we can fine grained

Thanks

**Comment by tomkerkhove** [comment]
@zroubalik @JorTurFer Do you know why this is the case? Likely this is not intentional

**Reply by JorTurFer** [reply]
I'd need to check it in depth but I guess that it's for allowing subresource/scale. Tbh, I haven't noticed it before and I don't like it because it enforces too much privileges.

Maybe could we generate theist dinamically based on given list of resources? Or at least a helm parameter with the list even though we use * as default, but it'd support that users customize the list with their custom resources

**Reply by tomkerkhove** [reply]
Isn't this managed by KEDA core?

**Reply by JorTurFer** [reply]
yes, but in core manifests we can't remove it or we will break the subresources/scale as we need to introspect them: https://github.com/kedacore/keda/blob/main/pkg/scaling/resolver/scale_resolvers.go#L116-L130
This is the line where the * are added: https://github.com/kedacore/keda/blob/main/controllers/keda/scaledobject_controller.go#L58
In helm, we could change it because it's dynamic, but in raw yaml we cannot do it, users have to update the yaml by themselves to limit the resources if they want (we should clarify how to do it)

**Reply by raorugan** [reply]
ok.. @JorTurFer can users change in the yaml - https://github.com/kedacore/charts/blob/v2.10.1/keda/templates/10-keda-clusterrole.yaml as per the limits identified for their resources similar to secrets?
![image](https://github.com/kedacore/keda/assets/45637559/b0456619-4243-4fb2-bed3-fddfdf24d4d0)

**Reply by JorTurFer** [reply]
currently they cant, but we can update the chart to allow it

**Reply by tomkerkhove** [reply]
I'd prefer not to deviate too much from Core manifests though. however, I get this can be a pain but let's try to reduce the skewing as much as possible.

**Reply by DilanUA** [reply]
Indeed it's too much privileges to allow read access for all resources in all api groups.
It can be considered even as a critical blocker for a production run from a security PoV. 
So cannot we be more specific in this case depending on the actual application use case?

**Reply by JorTurFer** [reply]
> I'd prefer not to deviate too much from Core manifests though. however, I get this can be a pain but let's try to reduce the skewing as much as possible.

Core manifests don't allow anymore, I mean, they are generated once based on the code and that's it. 
As I said, we could explain in docs how to update it to reduce the scope for yaml users and support limiting them throught helm, but I'd not reduce the scope by code because it's a huge breaking change

**Reply by zroubalik** [reply]
The reason is for KEDA to be able to scale any CustomResource. If you know what set of resources are you gonna scale, you can modify the role.

**Reply by JorTurFer** [reply]
So, should we document how to update the role for yaml users and support it in helm?

**Reply by zroubalik** [reply]
Yes, definitely!

**Reply by IsuruBoyagane15** [reply]
A quick check;
If I am going to scale only "deployment" resources using KEDA, replacing the rule with

```yaml
- apiGroups: ["apps"]
  resources: ["deployments"]
  verbs: ["get"]
  ```

 would be the solution, right?

**Reply by JorTurFer** [reply]
I think so

**Comment by joebowbeer** [comment]
@zroubalik @JorTurFer Any update on this?

I installed the helm chart v2.13.1 with all secret restrictions enabled

```yaml
  permissions:
    operator:
      restrict:
        secret: true
    metricServer:
      restrict:
        secret: true
```

but keda-operator still has "get" access to all secrets by way of this permissive rule in the keda-operator ClusterRole:

```yaml
- apiGroups:
  - '*'
  resources:
  - '*'
  verbs:
  - get
```

```bash
kubectl auth can-i get secrets --as=system:serviceaccount:keda:keda-operator
yes
```

I think the documentation for the restrict fields is misleading, as it implies that access to secrets will be (completely) restricted.

It would be helpful if the docs explained how to finish the job.

**Reply by JorTurFer** [reply]
```yaml
- apiGroups:
  - '*'
  resources:
  - '*'
  verbs:
  - get
```
This permission is required because KEDA can potentially work with any resource which implements /scale, so it needs that permission for working. The filter is done inside the code at some different levels:

- at logic level, we don't read the secret: https://github.com/kedacore/keda/blob/main/pkg/scaling/resolver/scale_resolvers.go#L402
- at client level, we register the secret lister namespace: https://github.com/kedacore/keda/blob/main/cmd/operator/main.go#L210

To avoid the * , you can update that permission adding the other permissions required for your workloads, if you just use `Deployments`, replacing it with 
```yaml
- apiGroups: ["apps"]
  resources: ["deployments"]
  verbs: ["get"]
```
 as suggested above (or any other workload resource)

Given that, I think that this can be a good improvement to docs, are you willing to open a PR clarifying it?

**Reply by joebowbeer** [reply]
I'm creating two issues:
1. https://github.com/kedacore/keda-docs/issues/1307
2. https://github.com/kedacore/charts/issues/605

**Reply by JorTurFer** [reply]
Let me move my comment there to track it too

**Comment by joebowbeer** [comment]
See recent changes https://github.com/kedacore/charts/pull/625
